### PR TITLE
Add new metadata prop to Slack send message

### PR DESCRIPTION
**Pull request by jonparker**
Add new metadata prop to Slack's `chat.postMessage` API as detailed in https://api.slack.com/metadata/publishing#sending_metadata

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

**Comment by dylburger** [comment]
Thank you so much for submitting this! We've added it to our backlog to review, and our team has been notified.

**Review by vellames-turing — DISMISSED** [review]

**Comment by dannyroosevelt** [comment]
I did a quick test and the new prop isn't working for me. I _think_ it likely has to do w/ the structure -- we have the prop as an `object` now, but it looks like Slack wants the an explicit key-value in the form of `event_type` (string) and `event_payload` (object).

**Comment by dannyroosevelt** [comment]
Oh also, we need to increment version patch for all components that reference the `send-message-common.mjs` file (`0.2.1` --> `0.2.2` for example).

**Comment by dannyroosevelt** [comment]
@jonparker FYI I took this out of the QA queue for now -- let me know if you're working on fixing the issues above or if you'd like some help! Once you're ready, we'll get this reviewed and tested asap.

File: components/slack/slack.app.mjs [hunk 1]
```diff
@@ -383,6 +383,18 @@ export default {
...
+      label: "Metadata Event Type",
+      description: "The name of the metadata event",
+      optional: true,
+    },
+    metadata_event_payload: {
+      type: "string",
+      label: "Metadata Event Payload",
+      description: "The payload of the metadata event",
+      optional: true,
+    },
```
Comment (jcortes) [inline components/slack/slack.app.mjs:397]: Quick question how are you sending this property as an object if you defined it as an string and I didn't see any parsing on this string either? thanks

**Review by jcortes — COMMENTED** [review]
Hi @jonparker I've just added a quick question before moving on with this PR! Thanks in advance!

File: components/slack/slack.app.mjs (hunk 1 above)
Comment (jonparker) [reply components/slack/slack.app.mjs:397]: I was trying to work out how to define it as an object and noticed the `blocks` prop is a string so decided to copy that.
But I think you are right. I'll have to check if the blocks are getting parsed.

What would you recommend?

File: components/slack/actions/send-message-common.mjs [hunk 2]
```diff
@@ -88,6 +100,13 @@ export default {
       blocks.push(sentViaPipedreamText);
     }
 
+    if (this.metadata_event_type) {
+      this.metadata = {
+        event_type: this.metadata_event_type,
+        event_payload: this.metadata_event_payload,
+      };
+    }
```
Comment (jcortes) [inline components/slack/actions/send-message-common.mjs:108]: I would add the parsing here like
```suggestion
    let metadataEventPayload;

    if (this.metadata_event_type) {
      try {
        metadataEventPayload = JSON.parse(this.metadata_event_payload);
      } catch (error) {
        throw new Error(`Invalid JSON in metadata_event_payload: ${error.message}`);
      }

      this.metadata = {
        event_type: this.metadata_event_type,
        event_payload: metadataEventPayload,
      };
    }
```

File: components/slack/slack.app.mjs [hunk 3]
```diff
@@ -383,6 +383,18 @@ export default {
...
+    metadata_event_type: {
+      type: "string",
+      label: "Metadata Event Type",
+      description: "The name of the metadata event",
+      optional: true,
+    },
+    metadata_event_payload: {
+      type: "string",
+      label: "Metadata Event Payload",
+      description: "The payload of the metadata event",
```
Comment (jcortes) [inline components/slack/slack.app.mjs:395]: And here I would specify how should be build the object as a string. Something like.
```suggestion
      description: "The payload of the metadata event. Must be a JSON string e.g. `{\"key\": \"value\"}`",
```

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @jonparker I've just made a couple of suggestions on how you can handle the parsing of the property you need to set.

**Comment by jonparker** [comment]
@jcortes I've accepted your suggestions.

Can it be merged now?

**Comment by dannyroosevelt** [comment]
> @jcortes I've accepted your suggestions.
> 
> Can it be merged now?

We'll get one final set of eyes on it then we should be good to ship it after that! (should be today most likely)

**Comment by dannyroosevelt** [comment]
We just shipped a number of unrelated updates to a lot of our Slack actions, so it looks like there are now some merge conflicts -- @michelle0927 are you able to help resolve? If they aren't too messy then we should be able to ship asap, since @jonparker already tested his specific feature end to end.

**Review by michelle0927 — APPROVED** [review]
Merge conflicts resolved!

**Comment by jonparker** [comment]
Thanks @michelle0927

**Comment by vunguyenhung** [comment]
Hello @jonparker @michelle0927, all test cases are passed! Ready for release! 

Test report
[Slack_Send_Metadata.pdf](https://github.com/PipedreamHQ/pipedream/files/8700785/Slack_Send_Metadata.pdf)


----
**Important Note**
At first, I've tried to test this new changes, but I can't seem to publish the `send-message-public-channel`. The error as below:
```
component create api call responded with status: 400, body: {"error":"{\"name\":\"UserError\",\"message\":\"ReferenceError: slack is not defined\"}"}
```

As I checked it's because the `send-message-public-channel` is using the `slack` variable that is not defined. It supposed to be `common.props.slack`, so I have pushed a commit to fix them all for affected actions

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**